### ISSUE TITLE
Authn is NOT Cheap, Provide a Request Limiter Filter before Authn

### ISSUE THREAD

**Issue by answer1991** [kubernetes/kubernetes#83745]
### Authn is NOT Cheap 

Authn is NOT cheap, both x509, ServiceAccount, Webhook and some other authn methods. Especially, before the #80465 merged, we found our API Server do ServiceAccount authn cost more then 5min when it's on heavy load, the logs like this:

```bash
I1010 09:50:37.727558       1 request.go:538] Throttling request took 4m11.678233345s, request: GET:https://127.0.0.1:6443/api/v1/namespaces/foo-ds/serviceaccounts/default
```

With [max-inflight Filter](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go) and [priority-and-fairness](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md), clients can also make API Server crash through keep sending their requests frequently, because the API Server is busy with requests' authn but the [max-inflight Filter](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go) and [priority-and-fairness](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md) have not worked.

We should provide a way to protect API Server from crash caused by clients' huge amount of requests. With this ability, API Server could reject requests according to clients' User-Agent, source or other something, but not the information got after authn.

### User Story

Our cluster has more than 8k+ Nodes, and 20+ DaemonSets, and 150k+ Pods, and there are lots of API Server extremely bad situation even crash cases happened in our cluster because of clients' huge amount of requests.

For now, when API Server is hit by a large number of requests and even Controller stop working, we have no method to recover our API Server, but stop our DaemonSets' container through Docker API . DaemonSets' Pod send too many requests, but we can not stop DaemonSets Pods, because Controller stop working when API Server is on heavy load.

### Limit Request by User-Agent

I recommend API Server provide an ability which can reject requests according to requests' User-Agent, this ability is really useful when API Server is in extremely bad situation.

In a production env, when our API Server is going to crash, the first thing we do is limit the specified User-Agent requests, and we get which User-Agents should be limited from our Grafana monitor graph.

Actually, client-go User-Agent information is very illustrative and it's can be trusted.

* Why User-Agent?

1. We can statistics and metric requests' User-Agent through Prom and monitor through Grafana. We can limit and reject requests from some User-Agent, which is really can protect API Server from crash when in emergency cases.
2. API Server can get User-Agent without authn, the ability can be implemented as a upper Filter before authn filter. API Server can reject requests with minimal cost.

**Comment by answer1991** [comment]
We can set [max-inflight Filter](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go) before authn filter, except one thing, max-inflight  should not limit the masters' request.

**Comment by liggitt** [comment]
note that https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md assumes availability of authentication-provided info:

> Each FlowSchema has:
> * A boolean test of an authenticated request;
> ...

> A FlowSchema’s boolean test is constructed from atomic tests. Each atomic test compares an authenticated request attribute --- selected from either the client identity attributes or those that characterize the work being requested --- with a literal value (scalar, pattern, or set).
> ...

**Comment by liggitt** [comment]
> Authn is NOT cheap, both x509 and ServiceAccount. Especially, before the #80465 merged, we found our API Server do ServiceAccount authn cost more then 5min when it's on heavy load

what version of the API server was that on? I'm curious what you see with the current state of the API server, with these three fixes especially:
1. API object lookup by the auth layer uses informer-fed caches first, and only falls back to live lookup if the cache lookup fails (https://github.com/kubernetes/kubernetes/pull/71816, merged in 1.14)
2. Thread context to client so that rate limiter tokens that would take longer to obtain than the context timeout return failure and don't block (https://github.com/kubernetes/kubernetes/pull/79375, merged in 1.16)
3. Drop QPS/Burst rate limiting for internal loopback clients (https://github.com/kubernetes/kubernetes/pull/80465, merged in 1.17, picked to 1.14/1.15/1.16)

**Comment by answer1991** [comment]
@liggitt thanks for your reply.

> note that priority-and-fairness assumes availability of authentication-provided info

Yes, we should consider priority-and-fairness can provide some ability to limit requests before authn, according User-Agent or other some information but not authn info.

>  I'm curious what you see with the current state of the API server

For now, we use v1.12 version, but we have cherry-picked these PRs except the first one.  I just worry about SA authn performance in extremely bad situation. Besides x509 and ServiceAccount authn, API Server has Webhook authn method and some other method, which is NOT cheap too.

I'd like to share our user story:

Our cluster has more than 8k+ Nodes, and 20+ DaemonSets, and 150k+ Pods, and there are lots of API Server extremely bad situation even crash cases happened in our cluster because of clients' huge amount of requests.

For now, when API Server is hit by a large number of requests and even Controller stop working, we have no method to recover our API Server, but stop our DaemonSets' container through Docker API . DaemonSets' Pod send too many requests, but we can not stop DaemonSets Pods, because Controller stop working when API Server is on heavy load.

**Comment by yue9944882** [comment]
> For now, we use v1.12 version

it reminds me of CVE-2018-16875. is the fix picked (v1.12.5+ by #72072)?

**Comment by answer1991** [comment]
@yue9944882 

> it reminds me of CVE-2018-16875. is the fix picked

Sure, we use golang 1.12.10 to compile our API Server.

**Comment by answer1991** [comment]
I had updated our user story.

**Comment by liggitt** [comment]
> we should consider priority-and-fairness can provide some ability to limit requests before authn, according User-Agent or other some information but not authn info.

User-Agent, etc, are not trustworthy sources to use to prioritize traffic

> we have cherry-picked these PRs except the first one

the first fix was the one that would improve service account authentication the most

**Comment by MikeSpreitzer** [comment]
Yes, there is a tension here between wanting to make decisions early and wanting to make decisions based on authenticated information.  There is value in doing some filtering before authentication, but that does not remove the value of also making decisions based on authenticated information.  For the early filtering, don't want to filter as early and cheaply as possible?  I suspect we need to go earlier and lower than is suggested here.

**Comment by liggitt** [comment]
It would be good to gather evidence of current service account authentication cost that makes it prohibitive to place before the priority-and-fairness filter. See also https://github.com/kubernetes/kubernetes/issues/83375

**Comment by answer1991** [comment]
@liggitt 

> the first fix was the one that would improve service account authentication the most

Yes, thanks for your 

> User-Agent, etc, are not trustworthy sources to use to prioritize traffic

Using User-Agent to limit requests is NOT prioritize traffic, but as a emergency strategy to protect API Server when it's in extremely bad situation.

And the filter User-Agent limiter can cooperate well with other limiters by authn info, such as max-inflight and priority-and-fairness, but NOT replace them.

**Comment by liggitt** [comment]
> Using User-Agent to limit requests is NOT prioritize traffic, but as a emergency strategy to protect API Server when it's in extremely bad situation.
> 
> And the filter User-Agent limiter can cooperate well with other limiters by authn info, such as max-inflight and priority-and-fairness, but NOT replace them.

The methods I can think of that rate limit specific callers based on untrusted/unverified data (like User-Agent) would allow a bad actor to starve a client that should be high priority or unlimited by flooding requests with a matching User-Agent.

**Comment by answer1991** [comment]
> There is value in doing some filtering before authentication, but that does not remove the value of also making decisions based on authenticated information.

Yes, I mean, these two limiter can cooperate well. 

And using User-Agent to limit requests is really a good emergency strategy to protect API Server when it's in extremely bad situation, as it's cheap. And we can decide immediately which User-Agent should be limited through our Grafana monitor graph.

**Comment by answer1991** [comment]
> The methods I can think of that rate limit specific callers based on untrusted/unverified data (like User-Agent) would allow a bad actor to starve a client that should be high priority or unlimited by flooding requests with a matching User-Agent.

In a production env, when our API Server is going to crash, the first thing we do is limit the specified User-Agent requests, and we get which User-Agents should be limited from our Grafana monitor graph.

Actually, client-go User-Agent information is very illustrative and it's can be trusted.

**Comment by liggitt** [comment]
> Actually, client-go User-Agent information is very illustrative and it's can be trusted.

For well-behaved clients, it could be useful, but if it's unverified, I don't see how it can be trusted.  How would you detect/prevent a bad actor from flooding requests with a user-agent matching a client that should be high priority or unlimited?

**Comment by mikedanese** [comment]
Since 1.12, authentication has become substantially cheaper after recent improvements we've made. From @liggitt's comment:

> * API object lookup by the auth layer uses informer-fed caches first, and only falls back to live lookup if the cache lookup fails (#71816, merged in 1.14)
> * Thread context to client so that rate limiter tokens that would take longer to obtain than the context timeout return failure and don't block (#79375, merged in 1.16)
> * Drop QPS/Burst rate limiting for internal loopback clients (#80465, merged in 1.17, picked to 1.14/1.15/1.16)

Additional improvements:

* We migrated off of the LRU cache which substantially improves cache hits. https://github.com/kubernetes/kubernetes/pull/84424
* We will be sharing token lookups after https://github.com/kubernetes/kubernetes/pull/85242

The cases between "we can't preform authentication before pushing back on clients" and "there's really nothing we can do in the kube-apiserver to keep up" are getting progressively slimmer. We should continue to work to make the strategy of key'ing priority and fairness off of the authenticated user. Can we reevaluate the claims with fixed versions of k8s and open up individual performance issues before proceeding?
